### PR TITLE
Improved the implementation of :meth:`.CoordinateSystem.get_area` to work without Riemann rectangles, added new :meth:`.CoordinateSystem.get_area_with_riemann_rectangles`

### PR DESCRIPTION
**Pull request by leleogere**
## Overview: What does this pull request change?

Modify the implementation of `get_area` to create a single Polygon instead of a brunch of Riemann rectangles.
The old method have been renamed `get_area_with_riemann_rectangles` (warning: breaking change)

## Motivation and Explanation: Why and how do your changes improve the library?

The performances were very badly impacted by the number of Mobjects on the scene, and it was very hard to animate an area in a reasonable amount of time.

## Links to added or changed documentation pages
 
[`get_area`](https://manimce--1990.org.readthedocs.build/en/1990/reference/manim.mobject.coordinate_systems.CoordinateSystem.html?highlight=get_area#manim.mobject.coordinate_systems.CoordinateSystem.get_area)
[`get_area_with_riemann_rectangles`](https://manimce--1990.org.readthedocs.build/en/1990/reference/manim.mobject.coordinate_systems.CoordinateSystem.html?highlight=get_area_with_riemann_rectangles#manim.mobject.coordinate_systems.CoordinateSystem.get_area_with_riemann_rectangles)

## Reviewer Checklist
- [ ] The PR title is descriptive enough for the changelog, and the PR is labeled correctly
- [ ] If applicable: newly added non-private functions and classes have a docstring including a short summary and a PARAMETERS section
- [ ] If applicable: newly added functions and classes are tested

### CODE REVIEW

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [inline manim/mobject/coordinate_systems.py:1148]: **Question :** Why is it reversed here ?

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [inline manim/mobject/coordinate_systems.py:1142]: **Question :** Why is this a large comparison, rather than a strict one ? (I mean, I guess it does not pose problem, just asking).

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [inline manim/mobject/coordinate_systems.py:1091]: **Issue** Seeing l.1138 it should not be optional and should be `Tuple[float, float`

File: tests/test_graphical_units/test_axes.py
Comment (GameDungeon) [inline tests/test_graphical_units/test_axes.py:175]: Could you add a test for the new version too?

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [inline manim/mobject/coordinate_systems.py:1094]: **Nitpick** : I don't really like the name `bounded`, since it often implied that the parameter is a boolean variable. Why not `bounded_func`, or something ?

File: manim/mobject/coordinate_systems.py
Comment (GameDungeon) [inline manim/mobject/coordinate_systems.py:1142]: You should use filter here instead of list comps

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]
Brilliant, I left some suggestions. 

I just have two remarks : The case where the function used as `bounded` is defined in a smaller interval than `graph` is not handled (a and b are not updated accordingly). I think it's an issue, since I would expect the area to be showed on the minimum shared interval. 
There is no support when `bounded` and `graph` have mismatching discontinuities (I think).
 
I would be nice you write some tests for these cases, if you decide that they need to be fixed. 

Thanks for the PR !

File: manim/mobject/coordinate_systems.py
Comment (GameDungeon) [inline manim/mobject/coordinate_systems.py:1147]: same as above

**Review by GameDungeon — CHANGES_REQUESTED** [review]
I have some small changes, but looks good

**Review by GameDungeon — CHANGES_REQUESTED** [review]
One other thing: Try to use self.points instead of get_points() when possible.

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [reply manim/mobject/coordinate_systems.py:1142]: I thought that as well, but at the end of the day it would be maybe a little less readable in my eyes.

File: manim/mobject/coordinate_systems.py
Comment (leleogere) [reply manim/mobject/coordinate_systems.py:1148]: The polygon follows the first curve, from the left to the right, and then "jumps" to the second curve, but it has to follow it from the right to the left to avoid crossing (like in a square jumping from the top left corner to the bottom left one)

File: manim/mobject/coordinate_systems.py
Comment (behackl) [reply manim/mobject/coordinate_systems.py:1142]: `filter` is not particularly Pythonic. If there is no good reason for using `filter` over the list comprehension, I'd rather keep this as-is.

Comment (GameDungeon) [reply manim/mobject/coordinate_systems.py:1142]: (I also think flake will wine about it, with the flake updates in an open pull that should be merged after the update goes out)

Comment (leleogere) [reply manim/mobject/coordinate_systems.py:1142]: I guess I want to include the two points on the right and the left in order to be exactly above (or below) the corresponding `self.c2p(a)` and self.c2p(b)` (not sure if that makes sense)

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [reply manim/mobject/coordinate_systems.py:1148]: Makes total sense. Thanks a lot !

File: manim/mobject/coordinate_systems.py
Comment (leleogere) [reply manim/mobject/coordinate_systems.py:1094]: I agree, I just used the same parameter as the original one. Should I modify the name of the original one too then?

File: tests/test_graphical_units/test_axes.py
Comment (leleogere) [reply tests/test_graphical_units/test_axes.py:175]: It should be just below that one

**Comment by behackl** [comment]
> One other thing: Try to use self.points instead of get_points() when possible.

The last time I've checked, using `self.points` directly breaks the metaclass compatibility and the Mobject will not work with OpenGL rendering. This might have changed in the meantime.

File: manim/mobject/coordinate_systems.py
Comment (GameDungeon) [reply manim/mobject/coordinate_systems.py:1142]: Ok then, I see this is a topic of hot debate.

**Comment by GameDungeon** [comment]
I've used it before and all tests passed completely (on a function with tests). The get_points() function is a function with a body of return points

File: manim/mobject/coordinate_systems.py
Comment (leleogere) [reply manim/mobject/coordinate_systems.py:1142]: Why so?

**Comment by behackl** [comment]
> I've used it before and all tests passed completely (on a function with tests). The get_points() function is a function with a body of return points

Yes, just checked – it changed with #1898. All good.

File: manim/mobject/coordinate_systems.py
Comment (GameDungeon) [reply manim/mobject/coordinate_systems.py:1142]: https://stackoverflow.com/questions/3013449/list-comprehension-vs-lambda-filter

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [reply manim/mobject/coordinate_systems.py:1094]: I think you should modify it

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [reply manim/mobject/coordinate_systems.py:1142]: yes, I get it. Thanks for the clarification !

**Review by GameDungeon — APPROVED** [review]
After the changes others have suggested LGTM :)

**Comment by leleogere** [comment]
I think I fixed all!

**Review by huguesdevimeux — CHANGES_REQUESTED** [review]

File: manim/mobject/coordinate_systems.py
Comment (huguesdevimeux) [reply manim/mobject/coordinate_systems.py:1091]:
```suggestion
        x_range: Optional[Tuple[float, float]] = None,
```

Comment (leleogere) [reply manim/mobject/coordinate_systems.py]: Oops sorry yes I only fixed the `Optional` part, didn't the the `Tuple` part!

**Review by huguesdevimeux — APPROVED** [review]

**Comment by Darylgolden** [comment]
I think it makes sense to adapt [this example](https://manimce--1990.org.readthedocs.build/en/1990/examples.html#graphareaplot) to show the differences in the two area methods. What do you think?

**Comment by leleogere** [comment]
> I think it makes sense to adapt [this example](https://manimce--1990.org.readthedocs.build/en/1990/examples.html#graphareaplot) to show the differences in the two area methods. What do you think?

You mean one area using the old method, and the other the new one? If so do you have a preference on which one should use the old method?

**Comment by Darylgolden** [comment]
> 
> 
> > I think it makes sense to adapt [this example](https://manimce--1990.org.readthedocs.build/en/1990/examples.html#graphareaplot) to show the differences in the two area methods. What do you think?
> 
> You mean one area using the old method, and the other the new one? If so do you have a preference on which one should use the old method?

Yeah. I think it makes sense for the leftmost one (on stable, the one where the rectangles are easily visible) to be the Riemann one and for the right to use the new method.

**Comment by leleogere** [comment]
> > > I think it makes sense to adapt [this example](https://manimce--1990.org.readthedocs.build/en/1990/examples.html#graphareaplot) to show the differences in the two area methods. What do you think?
> > 
> > 
> > You mean one area using the old method, and the other the new one? If so do you have a preference on which one should use the old method?
> 
> Yeah. I think it makes sense for the leftmost one (on stable, the one where the rectangles are easily visible) to be the Riemann one and for the right to use the new method.

But the rectangles won't be visible anyway will they? (except if I pass a sufficiently big `dx_scaling`, but for this point wouldn't directly use `get_riemann_rectangles be more suited?)
The difference between those two is in the implementation, the final result is almost the same.

**Comment by Darylgolden** [comment]
> But the rectangles won't be visible anyway will they? (except if I pass a sufficiently big `dx_scaling`, but for this point wouldn't directly use `get_riemann_rectangles be more suited?)
> The difference between those two is in the implementation, the final result is almost the same.

image

I mean to keep this one as it is (the rectangles are clearly visible)

**Comment by leleogere** [comment]
Ok It should be fixed, and the new image should look similar to the current one now

**Review by behackl — CHANGES_REQUESTED** [review]
The tests are currently failing.

**Comment by leleogere** [comment]
Indeed there were an hidden bug that were revealed by my latest commit. It should be fixed now!

**Comment by hydrobeam** [comment]
What if we did away with the old `get_riemann_rectangles`-based implementation? iirc get_area works the way it does because of a lack of a proper solution, so having `get_area_with_riemann_rectangles linger around` now that we have a real solution for how to show area feels a bit odd imo.

**Comment by Darylgolden** [comment]
> 
> 
> What if we did away with the old `get_riemann_rectangles`-based implementation? iirc get_area works the way it does because of a lack of a proper solution, so having `get_area_with_riemann_rectangles linger around` now that we have a real solution for how to show area feels a bit odd imo.

It's helpful for showing how Riemann sums work. I think we should keep it.

**Comment by hydrobeam** [comment]
> It's helpful for showing how Riemann sums work. I think we should keep it.

This is fair, but part of my reasoning is that's just `get_riemann_rectangles` with some default parameters passed in (and some that are currently not viewable without digging into the source). So I'm not sure that it has a place in the library with the introduction of this PR, since when porting over `GraphScene`, methods that were essentially other methods + defaults were also scrapped.

**Comment by leleogere** [comment]
> What if we did away with the old get_riemann_rectangles-based implementation?
To be honest I quite agree with you. I only left it because some people said that it was useful to them.
> It's helpful for showing how Riemann sums work. I think we should keep it.
But to keep it you could just use `get_riemann_rectangles` with enough rectangles, that would look like an area (+ you would get access to additional parameters that were not available in he previous `get_area` method)

**Review by behackl — APPROVED** [review]
This PR is fine and can be merged in principle, sorry for the delay.

However, I agree with the recently raised point: given that we still have `get_riemann_rectangles`, we could simply remove `get_area_with_riemann_rectangles` (this is a breaking change in any case, and whether people use `get_riemann_rectangles` or `get_area_with_riemann_rectangles` doesn't make too much of a difference, IMO).

What do you think, @leleogere?

**Comment by leleogere** [comment]
> This PR is fine and can be merged in principle, sorry for the delay.
> 
> However, I agree with the recently raised point: given that we still have `get_riemann_rectangles`, we could simply remove `get_area_with_riemann_rectangles` (this is a breaking change in any case, and whether people use `get_riemann_rectangles` or `get_area_with_riemann_rectangles` doesn't make too much of a difference, IMO).
> 
> What do you think, @leleogere?

I completely agree with you, I just kept the other method because some people asked to keep it, but I agree that the old behaviour can still be easily reproduced with `get_riemann_rectangles`.

**Comment by behackl** [comment]
Then let's do it; would you be okay with removing `get_area_with_riemann_rectangles` before we merge this?

**Comment by leleogere** [comment]
> Then let's do it; would you be okay with removing `get_area_with_riemann_rectangles` before we merge this?

Done!

**Pull request merged**
